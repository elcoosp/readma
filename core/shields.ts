import { paramCase } from "@wok/case"
import type { ReadmeTemplateArgs } from "./types.ts"
/** Shield definition, see [shields.io](https://shields.io/) */
export type Shield = {
  name: string
  refName: string
  linkUrl: string
  shieldUrl: string
  lang?: string
}

/** Create a {@link Shield} */
export function shield(
  name: string,
  linkUrl: string,
  shieldUrl: string,
  lang?: string,
): Shield {
  const refName = paramCase(name)
  return { name, refName, linkUrl, shieldUrl, lang }
}
/** Util to create a {@link Shield.shieldUrl}, defaulting to [shields.io](https://shields.io/) */
export const shieldUrl = (path: string, style?: string) =>
  path.startsWith("https://")
    ? `${path}&style=${style}`
    : `https://img.shields.io/${path}.svg?${style ? `style=${style}` : ""}`
/** Dynamic list of shields, with some depending on args */
export const shields = (
  {
    githubUsername,
    repoUrl,
    repoName,
    badgeStyle,
    commitActivityInterval,
    linkedinUsername,
    branch,
    language,
    vcsName,
    workspaceMember,
  }:
    & Pick<
      ReadmeTemplateArgs,
      | "repoName"
      | "githubUsername"
      | "linkedinUsername"
      | "badgeStyle"
      | "language"
      | "workspaceMember"
      | "vcsName"
    >
    & {
      repoUrl: string
      commitActivityInterval: "w"
      branch: string
    },
): Shield[] =>
  [
    ...(linkedinUsername
      ? [
        shield(
          "LinkedIn",
          `https://linkedin.com/in/${linkedinUsername}`,
          "badge/-LinkedIn-black",
        ),
      ]
      : []),
    ...(language === "rs"
      ? [
        shield(
          "Crates MSRV",
          `https://crates.io/crates/${repoName}`,
          `crates/msrv/${repoName}`,
        ),
      ]
      : []),
    ...(language === "ts" && workspaceMember
      ? [
        shield(
          "JSR version",
          `https://jsr.io/@${repoName}/${workspaceMember}`,
          `jsr/v/@${repoName}/${workspaceMember}`,
        ),
      ]
      : []),
    ...(language === "rs"
      ? [
        shield(
          "Crates.io version",
          `https://crates.io//${workspaceMember ? workspaceMember : repoName}`,
          `crates/v/${workspaceMember ? workspaceMember : repoName}`,
        ),
      ]
      : []),
    shield(
      "Codecov",
      `https://codecov.io/${vcsName}/${githubUsername}/${repoName}/tree/${branch}${
        workspaceMember ? `/${workspaceMember}` : ""
      }`,
      `codecov/c/${vcsName}/${githubUsername}/${repoName}/${branch}`,
    ),
    shield(
      "Contributors",
      `${repoUrl}/graphs/contributors`,
      `github/contributors/${githubUsername}/${repoName}`,
    ),
    shield(
      "Forks",
      `${repoUrl}/network/members`,
      `github/forks/${githubUsername}/${repoName}`,
    ),
    shield(
      "Stargazers",
      `${repoUrl}/stargazers`,
      `github/stars/${githubUsername}/${repoName}`,
    ),
    shield(
      "Commit activity",
      `${repoUrl}/commits`,
      `github/commit-activity/${commitActivityInterval}/${githubUsername}/${repoName}`,
    ),
    shield(
      "Branch action runs",
      `${repoUrl}/actions?query=${encodeURIComponent(`branch:${branch}`)}`,
      `github/check-runs/${githubUsername}/${repoName}/${branch}`,
    ),
    shield(
      "License",
      `${repoUrl}/blob/master/LICENSE.txt`,
      `github/license/${githubUsername}/${repoName}`,
    ),
  ].map((x) => ({ ...x, shieldUrl: shieldUrl(x.shieldUrl, badgeStyle) }))
/** Render {@link shields} as markdown, return markdown refs which declare urls and badges to put at the top of the document */
export const renderShields = (shields: Shield[]) => {
  const shieldsBadges = shields.map((s) =>
    `[![${s.name}][${s.refName}-shield]][${s.refName}-url]`
  ).join("\n")
  const shieldsRefs = shields.map((s) =>
    `[${s.refName}-shield]: ${s.shieldUrl}\n[${s.refName}-url]: ${s.linkUrl}`
  ).join("\n")
  return { shieldsBadges, shieldsRefs }
}
