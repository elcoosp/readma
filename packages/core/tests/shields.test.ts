import { assertSnapshot } from '@std/testing/snapshot'
import { shields } from '../src/shields.ts'
const shieldsConfig = {
  githubUsername: 'elcoosp',
  repoUrl: 'https://github.com/elcoosp/readma',
  repoName: 'readma',
  badgeStyle: 'for-the-badge',
  commitActivityInterval: 'w',
  linkedinUsername: undefined,
  branch: 'main',
  language: 'ts',
  vcsName: 'github',
  workspaceMember: undefined,
  packageRegistry: 'npm',
} as const
Deno.test('shields - ts', async (t) => {
  const actual = shields(shieldsConfig)
  await assertSnapshot(t, actual)
})

Deno.test('shields - rs', async (t) => {
  const actual = shields({ ...shieldsConfig, language: 'rs' })
  await assertSnapshot(t, actual)
})

Deno.test('shields - linkedinUsername', async (t) => {
  const actual = shields({ ...shieldsConfig, linkedinUsername: 'test' })
  await assertSnapshot(t, actual)
})

Deno.test('shields - packageRegistry', async (t) => {
  const jsr = shields({ ...shieldsConfig, packageRegistry: 'jsr' })
  await assertSnapshot(t, jsr)
  const cratesIo = shields({
    ...shieldsConfig,
    packageRegistry: 'crates.io',
    language: 'rs',
  })
  await assertSnapshot(t, cratesIo)
})
