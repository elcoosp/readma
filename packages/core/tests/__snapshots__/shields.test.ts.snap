export const snapshot = {};

snapshot[`shields - ts 1`] = `
[
  {
    linkUrl: "https://codecov.io/github/elcoosp/readma/tree/main",
    logo: "codecov",
    name: "Codecov",
    refName: "codecov",
    shieldUrl: "https://img.shields.io/codecov/c/github/elcoosp/readma/main.svg?style=for-the-badge&logo=codecov&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/graphs/contributors",
    logo: undefined,
    name: "Contributors",
    refName: "contributors",
    shieldUrl: "https://img.shields.io/github/contributors/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/network/members",
    logo: undefined,
    name: "Forks",
    refName: "forks",
    shieldUrl: "https://img.shields.io/github/forks/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/stargazers",
    logo: undefined,
    name: "Stargazers",
    refName: "stargazers",
    shieldUrl: "https://img.shields.io/github/stars/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/commits",
    logo: undefined,
    name: "Commit activity",
    refName: "commit-activity",
    shieldUrl: "https://img.shields.io/github/commit-activity/w/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/actions?query=branch%3Amain",
    logo: "githubactions",
    name: "Branch action runs",
    refName: "branch-action-runs",
    shieldUrl: "https://img.shields.io/github/check-runs/elcoosp/readma/main.svg?style=for-the-badge&logo=githubactions&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/blob/master/LICENSE.txt",
    logo: undefined,
    name: "License",
    refName: "license",
    shieldUrl: "https://img.shields.io/github/license/elcoosp/readma.svg?style=for-the-badge",
  },
]
`;

snapshot[`shields - rs 1`] = `
[
  {
    linkUrl: "https://crates.io/crates/readma",
    logo: "rust",
    name: "Crates MSRV",
    refName: "crates-msrv",
    shieldUrl: "https://img.shields.io/crates/msrv/readma.svg?style=for-the-badge&logo=rust&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://codecov.io/github/elcoosp/readma/tree/main",
    logo: "codecov",
    name: "Codecov",
    refName: "codecov",
    shieldUrl: "https://img.shields.io/codecov/c/github/elcoosp/readma/main.svg?style=for-the-badge&logo=codecov&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/graphs/contributors",
    logo: undefined,
    name: "Contributors",
    refName: "contributors",
    shieldUrl: "https://img.shields.io/github/contributors/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/network/members",
    logo: undefined,
    name: "Forks",
    refName: "forks",
    shieldUrl: "https://img.shields.io/github/forks/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/stargazers",
    logo: undefined,
    name: "Stargazers",
    refName: "stargazers",
    shieldUrl: "https://img.shields.io/github/stars/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/commits",
    logo: undefined,
    name: "Commit activity",
    refName: "commit-activity",
    shieldUrl: "https://img.shields.io/github/commit-activity/w/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/actions?query=branch%3Amain",
    logo: "githubactions",
    name: "Branch action runs",
    refName: "branch-action-runs",
    shieldUrl: "https://img.shields.io/github/check-runs/elcoosp/readma/main.svg?style=for-the-badge&logo=githubactions&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/blob/master/LICENSE.txt",
    logo: undefined,
    name: "License",
    refName: "license",
    shieldUrl: "https://img.shields.io/github/license/elcoosp/readma.svg?style=for-the-badge",
  },
]
`;

snapshot[`shields - linkedinUsername 1`] = `
[
  {
    linkUrl: "https://linkedin.com/in/test",
    logo: "linkedin",
    name: "LinkedIn",
    refName: "linked-in",
    shieldUrl: "https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://codecov.io/github/elcoosp/readma/tree/main",
    logo: "codecov",
    name: "Codecov",
    refName: "codecov",
    shieldUrl: "https://img.shields.io/codecov/c/github/elcoosp/readma/main.svg?style=for-the-badge&logo=codecov&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/graphs/contributors",
    logo: undefined,
    name: "Contributors",
    refName: "contributors",
    shieldUrl: "https://img.shields.io/github/contributors/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/network/members",
    logo: undefined,
    name: "Forks",
    refName: "forks",
    shieldUrl: "https://img.shields.io/github/forks/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/stargazers",
    logo: undefined,
    name: "Stargazers",
    refName: "stargazers",
    shieldUrl: "https://img.shields.io/github/stars/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/commits",
    logo: undefined,
    name: "Commit activity",
    refName: "commit-activity",
    shieldUrl: "https://img.shields.io/github/commit-activity/w/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/actions?query=branch%3Amain",
    logo: "githubactions",
    name: "Branch action runs",
    refName: "branch-action-runs",
    shieldUrl: "https://img.shields.io/github/check-runs/elcoosp/readma/main.svg?style=for-the-badge&logo=githubactions&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/blob/master/LICENSE.txt",
    logo: undefined,
    name: "License",
    refName: "license",
    shieldUrl: "https://img.shields.io/github/license/elcoosp/readma.svg?style=for-the-badge",
  },
]
`;

snapshot[`shields - packageRegistry 1`] = `
[
  {
    linkUrl: "https://codecov.io/github/elcoosp/readma/tree/main",
    logo: "codecov",
    name: "Codecov",
    refName: "codecov",
    shieldUrl: "https://img.shields.io/codecov/c/github/elcoosp/readma/main.svg?style=for-the-badge&logo=codecov&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/graphs/contributors",
    logo: undefined,
    name: "Contributors",
    refName: "contributors",
    shieldUrl: "https://img.shields.io/github/contributors/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/network/members",
    logo: undefined,
    name: "Forks",
    refName: "forks",
    shieldUrl: "https://img.shields.io/github/forks/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/stargazers",
    logo: undefined,
    name: "Stargazers",
    refName: "stargazers",
    shieldUrl: "https://img.shields.io/github/stars/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/commits",
    logo: undefined,
    name: "Commit activity",
    refName: "commit-activity",
    shieldUrl: "https://img.shields.io/github/commit-activity/w/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/actions?query=branch%3Amain",
    logo: "githubactions",
    name: "Branch action runs",
    refName: "branch-action-runs",
    shieldUrl: "https://img.shields.io/github/check-runs/elcoosp/readma/main.svg?style=for-the-badge&logo=githubactions&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/blob/master/LICENSE.txt",
    logo: undefined,
    name: "License",
    refName: "license",
    shieldUrl: "https://img.shields.io/github/license/elcoosp/readma.svg?style=for-the-badge",
  },
]
`;

snapshot[`shields - packageRegistry 2`] = `
[
  {
    linkUrl: "https://crates.io/crates/readma",
    logo: "rust",
    name: "Crates MSRV",
    refName: "crates-msrv",
    shieldUrl: "https://img.shields.io/crates/msrv/readma.svg?style=for-the-badge&logo=rust&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://crates.io/crates/readma",
    logo: "rust",
    name: "Crates.io version",
    refName: "crates-io-version",
    shieldUrl: "https://img.shields.io/crates/v/readma.svg?style=for-the-badge&logo=rust&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://codecov.io/github/elcoosp/readma/tree/main",
    logo: "codecov",
    name: "Codecov",
    refName: "codecov",
    shieldUrl: "https://img.shields.io/codecov/c/github/elcoosp/readma/main.svg?style=for-the-badge&logo=codecov&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/graphs/contributors",
    logo: undefined,
    name: "Contributors",
    refName: "contributors",
    shieldUrl: "https://img.shields.io/github/contributors/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/network/members",
    logo: undefined,
    name: "Forks",
    refName: "forks",
    shieldUrl: "https://img.shields.io/github/forks/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/stargazers",
    logo: undefined,
    name: "Stargazers",
    refName: "stargazers",
    shieldUrl: "https://img.shields.io/github/stars/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/commits",
    logo: undefined,
    name: "Commit activity",
    refName: "commit-activity",
    shieldUrl: "https://img.shields.io/github/commit-activity/w/elcoosp/readma.svg?style=for-the-badge",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/actions?query=branch%3Amain",
    logo: "githubactions",
    name: "Branch action runs",
    refName: "branch-action-runs",
    shieldUrl: "https://img.shields.io/github/check-runs/elcoosp/readma/main.svg?style=for-the-badge&logo=githubactions&logoColor=f5f5f5",
  },
  {
    linkUrl: "https://github.com/elcoosp/readma/blob/master/LICENSE.txt",
    logo: undefined,
    name: "License",
    refName: "license",
    shieldUrl: "https://img.shields.io/github/license/elcoosp/readma.svg?style=for-the-badge",
  },
]
`;
